File: mdcc/cuckoo_hash_table.rb
class CuckooHashTable
  TABLE_SIZE = 8
  NUM_TABLES = 2

  def initialize
    @tables = Array.new(NUM_TABLES) { Array.new(TABLE_SIZE, nil) }
  end

  def hash1(x)
    x % TABLE_SIZE
  end

  def hash2(x)
    (x / TABLE_SIZE) % TABLE_SIZE
  end

  def insert(x)
    i1 = hash1(x)
    i2 = hash2(x)

    if @tables[0][i1].nil?
      @tables[0][i1] = x
    elsif @tables[1][i2].nil?
      @tables[1][i2] = x
    else
      y = @tables[0][i1]
      @tables[0][i1] = x
      insert(y)
    end
  end

  def find(x)
    i1 = hash1(x)
    i2 = hash2(x)
    @tables[0][i1] == x || @tables[1][i2] == x
  end
end

File: epaxos/src/cuckoo_hash_table.rs
const TABLE_SIZE: usize = 8;
const NUM_TABLES: usize = 2;

fn hash1(x: usize) -> usize {
    x % TABLE_SIZE
}

fn hash2(x: usize) -> usize {
    (x / TABLE_SIZE) % TABLE_SIZE
}

struct CuckooHashTable {
    tables: [Vec<Option<usize>>; NUM_TABLES],
}

impl CuckooHashTable {
    fn new() -> CuckooHashTable {
        CuckooHashTable {
            tables: [
                vec![None; TABLE_SIZE],
                vec![None; TABLE_SIZE],
            ],
        }
    }

    fn insert(&mut self, x: usize) {
        let i1 = hash1(x);
        let i2 = hash2(x);

        if self.tables[0][i1].is_none() {
            self.tables[0][i1] = Some(x);
        } else if self.tables[1][i2].is_none() {
            self.tables[1][i2] = Some(x);
        } else {
            let y = self.tables[0][i1].unwrap();
            self.tables[0][i1] = Some(x);
            self.insert(y);
        }
    }

    fn find(&self, x: usize) -> bool {
        let i1 = hash1(x);
        let i2 = hash2(x);

        self.tables[0][i1].map_or(false, |y| y == x)
            || self.tables[1][i2].map_or(false, |y| y == x)
    }
}
